base/src/Data/Macaw/Syntax/SExpr.hs

```diff
@@ -0,0 +1,231 @@
+  , withPosFrom
+  ) where
+
+import           Control.Applicative ( empty, (<|>), Alternative )
+import           Control.Monad ( MonadPlus )
+import qualified Control.Monad.RWS as RWS
+import qualified Control.Monad.Reader as CMR
+import           Control.Monad.ST ( ST )
+import qualified Control.Monad.State as CMS
+import           Control.Monad.Trans ( lift )
+import qualified Data.ByteString as BS
+import qualified Data.Map as Map
+import qualified Data.Parameterized.NatRepr as PN
+import qualified Data.Parameterized.Nonce as PN
+import           Data.Parameterized.Some ( Some(..) )
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as TE
+import qualified Data.Text.Encoding.Error as TEE
+import           Numeric.Natural ( Natural )
+import qualified Text.Megaparsec as TM
+import qualified Text.Megaparsec.Char as TMC
+import qualified Text.Megaparsec.Char.Lexer as TMCL
+
+import qualified Data.Macaw.CFG as MC
+import qualified Data.Macaw.Memory as MM
+import           Data.Macaw.Syntax.Atom ( Atom )
+import qualified Data.Macaw.Types as MT
+
+data Position = Position { file :: FilePath, line :: {-# UNPACK #-} !TM.Pos, column :: {-# UNPACK #-} !TM.Pos }
+  deriving (Eq, Ord, Show)
+
+data Positioned a = Positioned { position :: {-# UNPACK #-} !Position, positioned :: !a }
+  deriving (Eq, Functor, Ord, Show)
+
+data Layer f a = List [f a] | Atom !a
+  deriving (Show, Functor, Eq, Ord)
+
+newtype Syntax a = Syntax { unSyntax :: Positioned (Layer Syntax a) }
+  deriving (Show, Functor, Eq, Ord)
+
+newtype Datum a = Datum { unDatum :: Layer Datum a }
+  deriving (Show, Functor, Eq)
+
+class Syntactic a b | a -> b where
+  syntaxE :: a -> Layer Syntax b
+
+instance Syntactic (Layer Syntax a) a where
+  syntaxE = id
+
+instance Syntactic (Syntax a) a where
+  syntaxE = positioned . unSyntax
+
+pattern A :: Syntactic a b => b -> a
+pattern A x <- (syntaxE -> Atom x)
+
+pattern L :: Syntactic a b => [Syntax b] -> a
+pattern L xs <- (syntaxE -> List xs)
+
+syntaxPos :: Syntax a -> Position
+syntaxPos = position . unSyntax
+
+withPosFrom :: Syntax a -> b -> Positioned b
+withPosFrom stx x = Positioned (syntaxPos stx) x
+
+-- | Low-level syntax errors embedded in megaparsec parse errors
+data MacawSyntaxError where
+  InvalidZeroWidthBV :: MacawSyntaxError
+  InvalidVectorPayload :: Syntax Atom -> MacawSyntaxError
+  InvalidType :: Syntax Atom -> MacawSyntaxError
+  InvalidStatement :: Syntax Atom -> MacawSyntaxError
+  InvalidEndianness :: Syntax Atom -> MacawSyntaxError
+  MissingRegisterDefinition :: Natural -> MacawSyntaxError
+  DuplicateRegisterDefinition :: Natural -> MacawSyntaxError
+  InvalidAddressWidth :: Syntax Atom -> MacawSyntaxError
+
+deriving instance Eq MacawSyntaxError
+deriving instance Ord MacawSyntaxError
+
+data ParserContext arch ids =
+  ParserContext { memory :: MM.Memory (MC.ArchAddrWidth arch)
+                , nonceGen :: PN.NonceGenerator (ST ids) ids
+                }
+
+data ParserState arch ids =
```

Comment:
```suggestion
newtype ParserState arch ids =
```

Reply:
I'll do that if the parser state doesn't gain any additional fields by the time I'm done with the implementation